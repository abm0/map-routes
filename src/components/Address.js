import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';

import AddressInput from './AddressInput';
import AddressList from './AddressList';

import { fetchAddressList, addPoint } from 'store/actionCreators';

import { generateAddressId } from 'helpers';

const AddressBlock = styled.div`
  z-index: 2;
  position: fixed;
  top: 20px;
  left: 20px;
`;

export class Address extends Component {

  state = {
    isAddressListVisible: false,
  }
  
  static propTypes = {
    addresses: PropTypes.array.isRequired,
    fetchAddressList: PropTypes.func.isRequired,
    addPoint: PropTypes.func.isRequired,
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.addresses.length && !prevProps.addresses.length) {
      this.showAddressList()
    }
  }

  onClickOutside = () => {
    this.hideAddressList();
  }

  showAddressList() {
    this.setState({ isAddressListVisible: true });
  }

  hideAddressList() {
    this.setState({ isAddressListVisible: false });
  }
  
  handleAddressInputKeyDown = (e) => {
    const { addresses } = this.props;

    if (!addresses.length) return;
    
    if (e.keyCode === 13) {
      this.props.addPoint(addresses[0], generateAddressId());
      this.hideAddressList();
    }
  }
  
  render() { 
    const {
      onClickOutside,
      handleAddressInputKeyDown,
    } = this;

    const { 
      addresses,
      fetchAddressList,
      addPoint,
      isAddressFetching,
    } = this.props;

    const {
      isAddressListVisible,
    } = this.state;

    return (
      <AddressBlock>
        <AddressInput
          fetchAddressList={fetchAddressList}
          onClickOutside={onClickOutside}
          isAddressFetching={isAddressFetching}
          onKeyDown={handleAddressInputKeyDown}
        />
        <AddressList
          addresses={addresses}
          isVisible={isAddressListVisible}
          addPoint={addPoint}
        />
      </AddressBlock>
    );
  }
}

const mapStateToProps = (state) => ({
  addresses: state.addresses.list,
  isAddressFetching: state.addresses.isFetching,
});

const mapActionCreators = {
  fetchAddressList,
  addPoint,
};
 
export default connect(mapStateToProps, mapActionCreators)(Address);