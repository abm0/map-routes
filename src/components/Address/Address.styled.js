import styled from 'styled-components';

export const AddressBlock = styled.div `
  z-index: 2;
  position: fixed;
  top: 20px;
  left: 20px;

  @media screen and (max-width: 448px) {
    top: 0;
    left: 0;
    width: 100%;
  }
`;
