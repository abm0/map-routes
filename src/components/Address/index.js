import React, { Component } from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';

import AddressInput from '../AddressInput';
import AddressList from '../AddressList';

import { AddressBlock } from './Address.styled';

import {
  fetchAddressList,
  fetchAddressListSuccess,
  addPoint,
} from 'store/actionCreators';

import {
  generateAddressId
} from 'helpers';


export class Address extends Component {

  state = {
    isAddressListVisible: false,
  }

  static propTypes = {
    addresses: PropTypes.array.isRequired,
    fetchAddressList: PropTypes.func.isRequired,
    fetchAddressListSuccess: PropTypes.func.isRequired,
    addPoint: PropTypes.func.isRequired,
  }

  componentDidUpdate(prevProps, prevState) {
    if (this.props.addresses.length && !prevProps.addresses.length) {
      this.showAddressList()
    }
  }

  onClickOutside = () => {
    this.hideAddressList();
  }

  showAddressList() {
    this.setState({
      isAddressListVisible: true
    });
  }

  hideAddressList() {
    this.setState({
      isAddressListVisible: false
    });
  }

  handleAddressInputKeyDown = (e) => {
    const {
      addresses
    } = this.props;

    if (!addresses.length) return;

    if (e.keyCode === 13) {
      this.props.addPoint(addresses[0], generateAddressId());
      this.hideAddressList();
    }
  }

  render() {
    const {
      onClickOutside,
      handleAddressInputKeyDown,
    } = this;

    const {
      fetchAddressList,
      fetchAddressListSuccess,
      addPoint,
      addresses,
      isAddressFetching,
    } = this.props;

    const {
      isAddressListVisible,
    } = this.state;

    return ( 
      <AddressBlock >
        <AddressInput
          fetchAddressList={fetchAddressList}
          fetchAddressListSuccess={fetchAddressListSuccess}
          onClickOutside={onClickOutside}
          isAddressFetching={isAddressFetching}
          onKeyDown={handleAddressInputKeyDown}
          isAddressListVisible={isAddressListVisible}
        /> 
        <AddressList 
          addresses={addresses}
          isVisible={isAddressListVisible}
          addPoint={addPoint} 
        /> 
      </AddressBlock>
    );
  }
}

const mapStateToProps = (state) => ({
  addresses: state.addresses.list,
  isAddressFetching: state.addresses.isFetching,
});

const mapActionCreators = {
  fetchAddressList,
  fetchAddressListSuccess,
  addPoint,
};

export default connect(mapStateToProps, mapActionCreators)(Address);