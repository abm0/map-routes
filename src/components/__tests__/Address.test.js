import React from 'react';
import { Address } from '../Address';
import puppeteer from 'puppeteer';
import ShallowRenderer from 'react-test-renderer/shallow';
import 'jest-styled-components';

const renderer = new ShallowRenderer();

describe('Address block', () => {
  let browser;
  let page;
  const width = 1920;
  const height = 1080;

  const APP_URL = 'http://localhost:3000/';
  
  it('should match the snapshot', () => {
    renderer.render(
      <Address
        addresses={[]}
        fetchAddressList={() => {}}
        addPoint={() => {}}
        isAddressFetching={false}
      />
    );

    const address = renderer.getRenderOutput();
    
    expect(address).toMatchSnapshot();
  });

  describe('in browser', () => {
    const addressInputSelector = '[data-e2e-id="address-input"]';
    const addressListItemSelector = '[data-e2e-id="address-item"]';
    
    beforeAll(async () => {
      browser = await puppeteer.launch({
        headless: false,
        slowMo: 80,
        args: [`--window-size=${width},${height}`],
      });
      page = await browser.newPage();
      await page.setViewport({ width, height });
      await page.goto(APP_URL);
    });

    afterAll(() => {
      browser.close();
    });
    
    it('should contain address input', async () => {
      const input = page.$eval(addressInputSelector, (el) => !!el);

      expect(input).toBeTruthy();
    });
    
    it('should show address list on request and add address to points list', async done => {      
      page.on('response', async (response) => {
        if (response.url().includes('/addresses')) {
          const addressItems = await page.$$(addressListItemSelector);

          expect(addressItems.length).toBeGreaterThan(0);

          done();
        }
      });
      
      await page.waitForSelector(addressInputSelector);
      await page.click(addressInputSelector);
      await page.type(addressInputSelector, 'Ульяновск');
      await page.waitForNavigation({waitUntil: "networkidle0"});
    }, 15000);
  });
});