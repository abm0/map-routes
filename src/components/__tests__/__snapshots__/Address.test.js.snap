// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Address block should match the snapshot 1`] = `
<ForwardRef(styled.div)>
  <AddressInput
    fetchAddressList={[Function]}
    isAddressFetching={false}
    onClickOutside={[Function]}
    onKeyDown={[Function]}
  />
  <UNDEFINED
    addPoint={[Function]}
    addresses={Array []}
    isVisible={false}
  />
</ForwardRef(styled.div)>
`;
